core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -221,6 +221,175 @@ private[spark] class BlockManager(
     new BlockManager.RemoteBlockDownloadFileManager(this)
   private val maxRemoteBlockToMem = conf.get(config.MAX_REMOTE_BLOCK_SIZE_FETCH_TO_MEM)
 
+  private abstract class BlockStoreUpdater[T](
+      blockSize: Long,
+      blockId: BlockId,
+      level: StorageLevel,
+      classTag: ClassTag[T],
+      tellMaster: Boolean,
+      keepReadLock: Boolean) {
+
+    protected def byteBuffer: ChunkedByteBuffer
+
+    protected def saveToDiskStore(): Unit
+
+    private def saveToMemoryStore(): Boolean = {
+      if (level.deserialized) {
+        val values =
+          serializerManager.dataDeserializeStream(blockId, byteBuffer.toInputStream())(classTag)
+        memoryStore.putIteratorAsValues(blockId, values, classTag) match {
+          case Right(_) => true
+          case Left(iter) =>
+            // If putting deserialized values in memory failed, we will put the bytes directly
+            // to disk, so we don't need this iterator and can close it to free resources
+            // earlier.
+            iter.close()
+            false
+        }
+      } else {
+        val memoryMode = level.memoryMode
+        memoryStore.putBytes(blockId, blockSize, memoryMode, () => {
+          if (memoryMode == MemoryMode.OFF_HEAP &&
+            byteBuffer.chunks.exists(buffer => !buffer.isDirect)) {
+            byteBuffer.copy(Platform.allocateDirectBuffer)
+          } else {
+            byteBuffer
+          }
+        })
+      }
+    }
+
+    /**
+     * Put the given data according to the given level in one of the block stores, replicating
+     * the values if necessary.
+     *
+     * If the block already exists, this method will not overwrite it.
+     *
+     * If keepReadLock is true, this method will hold the read lock when it returns (even if the
+     * block already exists). If false, this method will hold no locks when it returns.
+     *
+     * @return true if the block was already present or if the put succeeded, false otherwise.
+     */
+     def save(): Boolean = {
+      doPut(blockId, level, classTag, tellMaster, keepReadLock) { info =>
+        val startTimeMs = System.currentTimeMillis
+        // Since we're storing bytes, initiate the replication before storing them locally.
+        // This is faster as data is already serialized and ready to send.
+        val replicationFuture = if (level.replication > 1) {
+          Future {
+            // This is a blocking action and should run in futureExecutionContext which is a cached
+            // thread pool. The ByteBufferBlockData wrapper is not disposed of to avoid releasing
+            // buffers that are owned by the caller.
+            replicate(blockId, new ByteBufferBlockData(byteBuffer, false), level, classTag)
```

Comment:
Here the lazy val `TempFileBasedBlockStoreUpdater.bytes` is accessed from a different thread. 
But it seams to me lazy vals are thread safe based on this: [New representation and locking scope for local lazy vals](
https://www.scala-lang.org/news/2.12.0/#new-representation-and-locking-scope-for-local-lazy-vals)